In [ ]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import warnings

warnings.filterwarnings('ignore')

In [ ]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'lifeexpectancy'
table_name = 'lifeexpectancy'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

life_exp_df = pd.read_sql_query('select * from lifeexpectancy',con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

To complete this challenge, submit a Jupyter notebook containing your solutions to the following tasks. Moreover, you can also submit a kernel to the Kaggle containing your solutions.

Your goal in this challenge is to find the factors that affect the life expectancy. Specifically, you need to find out which factors increase the expected life in the countries and which factors decrease it.

First, load the dataset.

Detect the problems with the data such as missing values and outliers. Are there any nonsense values that seem to be stemmed from the data collection? For the missing values, discuss which technique would be the most suitable one in filling out these values. Regarding the outliers, discuss their potential effects on your analysis and select an appropriate method to deal with them.

Explore the data using univariate and multivariate exploration techniques. You should pay special attention to your target variable. In this regard, your focus should be on finding the relevant variables that may affect life expectancy.

In the feature engineering step, you need to select a suite of variables that you think would be ideal in the modeling phase. More concretely, you may discard some variables that are very correlated with the other ones or the variables that you think irrelevant with the life expectancy.

Summarize your findings. One of the most important skills of a data scientist is to convey ideas and findings to nontechnical people using understandable language. In this regard, one of the most effective ways to communicate your ideas is to do it using effective visualization.

Answer the following questions as you work through this challenge, then submit a link to your notebook.  

In [ ]:
life_exp_df.info()
life_exp_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

Country  Year  ... Income composition of resources  Schooling
0  Afghanistan  2015  ...                           0.479       10.1
1  Afghanistan  2014  ...                           0.476       10.0
2  Afghanistan  2013  ...                           0.470        9.9
3  Afghanistan  2012  ...                           0.463        9.8
4  Afghanistan  2011  ...                           0.454        9.5

[5 rows x 22 columns]

In [ ]:
#inspect columns names
life_exp_df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

In [ ]:
#clean column names
orig_cols = list(life_exp_df.columns)
new_cols = []
for col in orig_cols:
    new_cols.append(col.strip().replace(' ', '_').lower())
life_exp_df.columns = new_cols
life_exp_df.columns


Index(['country', 'year', 'status', 'life_expectancy', 'adult_mortality',
       'infant_deaths', 'alcohol', 'percentage_expenditure', 'hepatitis_b',
       'measles', 'bmi', 'under-five_deaths', 'polio', 'total_expenditure',
       'diphtheria', 'hiv/aids', 'gdp', 'population', 'thinness__1-19_years',
       'thinness_5-9_years', 'income_composition_of_resources', 'schooling'],
      dtype='object')

In [ ]:
def missingness_summary(df,print_log,sort):
  percentage_null = df.isnull().sum()*100/df.isnull().count()
  if sort == 'ascending':
    return percentage_null.sort_values()
  elif sort == 'descending':
    return percentage_null.sort_values(ascending=False)
  else: 
    return percentage_null

print(missingness_summary(life_exp_df,True,'descending'))

population                         22.191967
hepatitis_b                        18.822328
gdp                                15.248468
total_expenditure                   7.692308
alcohol                             6.603131
income_composition_of_resources     5.684139
schooling                           5.547992
bmi                                 1.157250
thinness__1-19_years                1.157250
thinness_5-9_years                  1.157250
diphtheria                          0.646698
polio                               0.646698
adult_mortality                     0.340368
life_expectancy                     0.340368
under-five_deaths                   0.000000
hiv/aids                            0.000000
measles                             0.000000
percentage_expenditure              0.000000
infant_deaths                       0.000000
status                              0.000000
year                                0.000000
country                             0.000000
dtype: flo

We see that population, hepatitis_b, and gdp are missing 22.2%, 18.8%, and 15.2% of data from those columns respectively. This percentage of data seems manageable. Let's explore the values further in these columns to see if we acutally have more nulls than appear. 

In [ ]:
life_exp_df.describe()

year  ...    schooling
count  2938.000000  ...  2775.000000
mean   2007.518720  ...    11.992793
std       4.613841  ...     3.358920
min    2000.000000  ...     0.000000
25%    2004.000000  ...    10.100000
50%    2008.000000  ...    12.300000
75%    2012.000000  ...    14.300000
max    2015.000000  ...    20.700000

[8 rows x 20 columns]

There are some strange values in our dataset. For example, BMI of 1 and BMI of 87 seem like nearly impossible values. A population of 34 doesn't make no sense. Infant deaths and under-5_deaths of 0 seem very unlikely. We will treat these unlikely values as true nulls. 

In [ ]:
life_exp_df['infant_deaths'] = life_exp_df['infant_deaths'].replace(0, np.nan)
life_exp_df['under-five_deaths'] = life_exp_df['under-five_deaths'].replace(0, np.nan)
#bmi seems to range from 20-40 in the 5th-95th percentile. We can safely say <5 or >60 should be nulls
life_exp_df.bmi = life_exp_df.apply(lambda x: np.nan if (x.bmi < 10 or x.bmi > 50) else x.bmi, axis=1)
# smallest pop is the world is 1000, so we can safely say anything pop under 500 is null
life_exp_df.population = life_exp_df.apply(lambda x: np.nan if x.population < 500 else x.population, axis=1)


life_exp_df.describe()

year  ...    schooling
count  2938.000000  ...  2775.000000
mean   2007.518720  ...    11.992793
std       4.613841  ...     3.358920
min    2000.000000  ...     0.000000
25%    2004.000000  ...    10.100000
50%    2008.000000  ...    12.300000
75%    2012.000000  ...    14.300000
max    2015.000000  ...    20.700000

[8 rows x 20 columns]

Now that we have dealt with some nonsense values, let's look again at the missing values from our dataset. 

In [ ]:
print(missingness_summary(life_exp_df,True,'descending'))

bmi                                49.557522
infant_deaths                      28.863172
under-five_deaths                  26.718856
population                         23.110960
hepatitis_b                        18.822328
gdp                                15.248468
total_expenditure                   7.692308
alcohol                             6.603131
income_composition_of_resources     5.684139
schooling                           5.547992
thinness__1-19_years                1.157250
thinness_5-9_years                  1.157250
polio                               0.646698
diphtheria                          0.646698
adult_mortality                     0.340368
life_expectancy                     0.340368
measles                             0.000000
hiv/aids                            0.000000
percentage_expenditure              0.000000
status                              0.000000
year                                0.000000
country                             0.000000
dtype: flo

Next lets try to come up with a sensible technique to fill these missing values

In [ ]:
life_exp_df.head(50)

country  year  ... income_composition_of_resources  schooling
0     Afghanistan  2015  ...                           0.479       10.1
1     Afghanistan  2014  ...                           0.476       10.0
2     Afghanistan  2013  ...                           0.470        9.9
3     Afghanistan  2012  ...                           0.463        9.8
4     Afghanistan  2011  ...                           0.454        9.5
5     Afghanistan  2010  ...                           0.448        9.2
6     Afghanistan  2009  ...                           0.434        8.9
7     Afghanistan  2008  ...                           0.433        8.7
8     Afghanistan  2007  ...                           0.415        8.4
9     Afghanistan  2006  ...                           0.405        8.1
10    Afghanistan  2005  ...                           0.396        7.9
11    Afghanistan  2004  ...                           0.381        6.8
12    Afghanistan  2003  ...                           0.373        6.5
13    Afghanistan  2002  ...                           0.341        6.2
14    Afghanistan  2001  ...                           0.340        5.9
15    Afghanistan  2000  ...                           0.338        5.5
16        Albania  2015  ...                           0.762       14.2
17        Albania  2014  ...                           0.761       14.2
18        Albania  2013  ...                           0.759       14.2
19        Albania  2012  ...                           0.752       14.2
20        Albania  2011  ...                           0.738       13.3
21        Albania  2010  ...                           0.725       12.5
22        Albania  2009  ...                           0.721       12.2
23        Albania  2008  ...                           0.713       12.0
24        Albania  2007  ...                           0.703       11.6
25        Albania  2006  ...                           0.696       11.4
26        Albania  2005  ...                           0.685       10.8
27        Albania  2004  ...                           0.681       10.9
28        Albania  2003  ...                           0.674       10.7
29        Albania  2002  ...                           0.670       10.7
30        Albania  2001  ...                           0.662       10.6
31        Albania  2000  ...                           0.656       10.7
32        Algeria  2015  ...                           0.743       14.4
33        Algeria  2014  ...                           0.741       14.4
34        Algeria  2013  ...                           0.737       14.4
35        Algeria  2012  ...                           0.732       14.4
36        Algeria  2011  ...                           0.724       14.0
37  Côte d'Ivoire  2001  ...                             NaN        NaN
38        Algeria  2010  ...                           0.714       13.6
39        Algeria  2009  ...                           0.705       13.1
40        Algeria  2008  ...                           0.697       12.6
41        Algeria  2007  ...                           0.690       12.3
42        Algeria  2006  ...                           0.686       12.3
43        Algeria  2005  ...                           0.680       12.0
44        Algeria  2004  ...                           0.673       11.7
45        Algeria  2003  ...                           0.663       11.5
46        Algeria  2002  ...                           0.653       11.1
47        Algeria  2001  ...                           0.644       10.9
48        Algeria  2000  ...                           0.636       10.7
49         Angola  2015  ...                           0.531       11.4

[50 rows x 22 columns]

Since our data appears to be in indexed by country and also indexed by year we can interpolate all the missing values. 